sw/airborne/modules/flight_plan_guided/flight_plan_guided.c

```diff
@@ -0,0 +1,427 @@
+
+  return false;
+}
+
+/**
+ * WaitUntilSpeedOrAltitude: Blocks while waiting until a certain altitude or speed is reached
+ * @param[in] fail_altitude, wait for altitude in [m] (positive z-axis is up)
+ * @param[in] speed, speed in [m/s] (positive z-axis is up)
+ * @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ */
+bool WaitUntilSpeedOrAltitude(float speed, float fail_altitude)
+{
+  if (autopilot.mode != AP_MODE_GUIDED) { return true; }
+
+  if (stateGetPositionEnu_f()->z > -fail_altitude) { return false; }
+  if (stateGetSpeedEnu_f()->z < -speed) { return true; }
+
+  return false;
+}
+
+/**
+ * WaitforHeading: Blocks untill wanted heading is achieved
+ * @param[in] heading, waits untill MAV reached certain heading [rad] (clockwise is positive)
+ *  @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ *
+ * TODO: Wrap the angle around -pi and pi
+ */
+bool WaitforHeading(float heading)
+{
+  if (fabs(heading - stateGetNedToBodyEulers_f()->psi) < 0.1) {
+    return false;
+  }
+  return true;
+}
+
+/**
+ * WaitForHMode: Blocks until specified horizontal mode is achieved
+ * @param[in] mode, number of the specific horizontal mode that you want to wait for
+ * @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ */
+bool WaitForHMode(uint8_t mode)
+{
+  if (guidance_h.mode == mode) {
+    return false;
+  }
+  return true;
+}
+
+
+/********************************************//**
+ *  Timer functions
+ ***********************************************/
+
+/**
+ * ResetSpecialTimer: Reset a special timer, that works outside of the block and stage time
+ * @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ */
+float specialtimer = 0;
+bool ResetSpecialTimer(void)
+{
+  specialtimer = 0;
+  return false;
+}
+
+/**
+ * WaitUntilTimer: Blocks while waiting until until the special timer reaches the specified amount of seconds
+ * @param[in] sec, wait for amount of seconds [sec]
+ * @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ */
+bool WaitUntilTimer(float sec)
+{
+  if (autopilot.mode != AP_MODE_GUIDED) { return true; }
+
+  specialtimer += 1.0f / ((float)NAV_FREQ);
+  if (specialtimer < sec) { return true; }
+
+  return false;
+}
+
+/**
+ * WaitUntilTimerOrAltitude: Blocks while waiting until the timer reaches the specified amount of seconds or an altitude has been reached
+ * @param[in] sec, wait for amount of seconds [sec]
+ * @param[in] fail_altitude, wait for altitude in [m] (positive z-axis is up)
+ * @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ */
+bool WaitUntilTimerOrAltitude(float sec, float fail_altitude)
+{
+  if (autopilot.mode != AP_MODE_GUIDED) { return true; }
+
+  if (stateGetPositionEnu_f()->z > fail_altitude) { return false; }
```

Comment:
Actually, this suggestion will cause the function to constantly return true if it has not passed the altitude (so the function will keep on running), which means it will not hit the timer check in the mean time. That is not what I want with this function